pkg/discovery/discovery.go

```diff
@@ -95,19 +98,30 @@ func (td *tidbDiscovery) Discover(advertisePeerUrl string) (string, error) {
 	currentCluster = td.clusters[keyName]
 	currentCluster.peers[podName] = struct{}{}
 
-	if len(currentCluster.peers) == int(replicas) {
+	if len(currentCluster.peers) == int(replicas) && tc.Spec.Cluster == nil {
 		delete(currentCluster.peers, podName)
 		return fmt.Sprintf("--initial-cluster=%s=%s://%s", podName, tc.Scheme(), advertisePeerUrl), nil
 	}
 
-	var pdClient pdapi.PDClient
-	if tc.IsHeterogeneous() {
-		pdClient = td.pdControl.GetPDClient(pdapi.Namespace(tc.GetNamespace()), tc.Spec.Cluster.Name, tc.IsTLSClusterEnabled())
-	} else {
-		pdClient = td.pdControl.GetPDClient(pdapi.Namespace(tc.GetNamespace()), tc.GetName(), tc.IsTLSClusterEnabled())
+	var pdClients []pdapi.PDClient
+	if tc.Spec.Cluster != nil && len(tc.Spec.Cluster.Name) > 0 {
+		namespace := tc.Spec.Cluster.Namespace
+		if len(namespace) == 0 {
+			namespace = tc.GetNamespace()
+		}
+		pdClients = append(pdClients, td.pdControl.GetClusterRefPDClient(pdapi.Namespace(namespace), tc.Spec.Cluster.Name, tc.Spec.Cluster.ClusterDomain, tc.IsTLSClusterEnabled()))
+	}
+	if tc.Spec.PD != nil {
+		pdClients = append(pdClients, td.pdControl.GetPDClient(pdapi.Namespace(tc.GetNamespace()), tc.GetName(), tc.IsTLSClusterEnabled()))
 	}
 
-	membersInfo, err := pdClient.GetMembers()
+	var membersInfo *pdapi.MembersInfo
+	for _, client := range pdClients {
+		membersInfo, err = client.GetMembers()
```

Comment:
Do we only need one `membersInfo` or combined `membersInfo`? If `len(pdClients)>1` `membersInfo` will be rewritten.

Reply:
如果第一个 pdClient 获得 member 成功了，循环会 break 🤔

Reply:
I see... Then why do we use a `pdClient` array? Why do we use another `pdClient` to do a retry instead of returning an error directly when we fail to `client.GetMembers()`?

Reply:
比如当前集群所有的PD都跪掉了，failover 需要拉起新的 pd，这个时候需要 discovery server 依然能工作来返回 pd 的启动参数。 🤔 所以尝试使用另外一个集群的 PD 来获取 member 信息了

Reply:
其实从 tc.Status.PD.PeerMembers 里面来拿更好一点，🤔 因为如果多个集群部署的话 🤔

Reply:
不过这个值也可能拿不到 🤔

Reply:
I get it.